### ISSUE TITLE
How much does fake_relu matter?

### ISSUE THREAD

**Issue by zjysteven** [MadryLab/robust_representations#10]
Hi,

Thanks for this awesome work. I just notice that in the `inversion_loss` function for representation inversion,
```python
def inversion_loss(model, inp, targ):
    _, rep = model(inp, with_latent=True, fake_relu=True)
    loss = ch.div(ch.norm(rep - targ, dim=1), ch.norm(targ, dim=1))
    return loss, None
```
the `fake_relu` is set to True. According to the source code of robustness library, this would make the gradients pass straight through during the backprop.

Then my question is as follows:
1) My best guess for using this `fake_relu` is that it can enable a better inversion process. But I have no idea how much it matters and I do not see discussions/instructions on this detail in your paper. Is it a common practice when doing representation inversion?

2) I also do not see why this `fake_relu` is only applied to the last "layer" in the ResNet architecture.
```python
    def forward(self, x, with_latent=False, fake_relu=False, no_relu=False):
        assert (not no_relu),  \
            "no_relu not yet supported for this architecture"
        out = F.relu(self.bn1(self.conv1(x)))
        out = self.layer1(out)
        out = self.layer2(out)
        out = self.layer3(out)
        out = self.layer4(out, fake_relu=fake_relu)
```

Any clarifications or explanations are appreciated!
Thanks

**Comment by dtsip** [comment]
Hi,

Thank you for your kind words.

So the issue that we encountered is: if a specific ReLU is 0 (at the random init or sometime later) then the inversion process will not try to match this neuron since no gradient is flowing backwards.

1. I do not remember how important this was for inversion, but it was definitely important for neuron visualization.
2. Yes, one could imagine applying fake ReLU throughout the network. But we found that only applying them to the last layer was good enough and did not experiment further.

Let us know if you have any further questions.

**Comment by zjysteven** [comment]
@dtsip Thanks for the response, and it makes things clear! 

One extra question though, is there any particular reason for using L2 norm bound (for both adversarial training and inversion)? I'm wondering this since recently when I'm trying to construct a "robust dataset" using Linf robust model and train another Linf robust model using this robust dataset, I can hardly get any nontrivial Linf robustness. I know things can be different between two norm bounds, but I'm not sure whether it is I'm doing something wrong or it is just things do not work in the Linf case. Do you have any thoughts on this? Thanks!

**Comment by dtsip** [comment]
Hm, interesting. We did not have any particular reason for choosing L2. I think our reasoning was that L2 is a tighter constraint (Linf perturbations would typically result in most pixels changing by the maximum allowed value). We also found in our [previous work](https://arxiv.org/abs/1805.12152) that L2 robust models tend to produce slightly cleaner large-epsilon perturbations.

In terms of Linf experiments, we did verify that properties such as feature visualization also hold there, but we did not run a full range of experiments. I am pretty sure that we did not try the "robustification" experiments though.

That being said, there is no reason why the same principle would not apply for Linf. What is the Linf bound that you are considering? Note that our L2 robustification experiments was 0.5 which is relatively small. For Linf I would try something like 1/255?
